src/main/java/org/jenkinsci/plugins/workflow/cps/LoggingInvoker.java

```diff
@@ -0,0 +1,159 @@
+ */
+final class LoggingInvoker implements Invoker {
+
+    public static Invoker create(boolean sandbox) {
+        return new LoggingInvoker(sandbox ? new SandboxInvoker() : new DefaultInvoker());
+    }
+
+    private final Invoker delegate;
+
+    private LoggingInvoker(Invoker delegate) {
+        this.delegate = delegate;
+    }
+
+    private void record(String call) {
+        CpsThreadGroup g = CpsThreadGroup.current();
+        if (g == null) {
+            assert false : "should never happen";
+            return;
+        }
+        CpsFlowExecution execution = g.getExecution();
+        if (execution == null) {
+            assert false : "should never happen";
+            return;
+        }
+        execution.recordInternalCall(call);
+    }
+
+    private static boolean isInternal(Class<?> clazz) {
+        if (clazz == Safepoint.class || clazz == CpsClosure2.class) {
+            return false;
+        }
+        if (clazz.getClassLoader() instanceof GroovyClassLoader) { // similar to GroovyClassLoaderWhitelist
+            return false;
+        }
+        // TODO more precise would be the logic in jenkins.security.ClassFilterImpl.isLocationWhitelisted
+        // (simply checking whether the class loader can “see”, say, jenkins/model/Jenkins.class
+        // would falsely mark third-party libs bundled in Jenkins plugins)
+        String name = clazz.getName();
+        if (name.startsWith("com.cloudbees.groovy.cps.")) {
+            // Likely synthetic call, as to CpsDefaultGroovyMethods.
+            return false;
+        }
+        // acc. to `find …/jenkinsci/*/src/main/java -type f -exec egrep -h '^package ' {} \; | sort | uniq` this is decent
+        return name.contains("jenkins") || name.contains("hudson") || name.contains("cloudbees");
+    }
+
+    private void maybeRecord(Class<?> clazz, Supplier<String> message) {
+        if (isInternal(clazz)) {
+            record(message.get());
+        }
+    }
+
+    private static @Nonnull Class<?> classOf(@CheckForNull Object receiver) {
+        if (receiver == null) {
+            return Void.class;
+        } else if (receiver instanceof Class) {
+            return (Class<?>) receiver;
+        } else {
+            return receiver.getClass();
+        }
+    }
+
+    @Override public Object methodCall(Object receiver, String method, Object[] args) throws Throwable {
+        Class<?> clazz = classOf(receiver);
+        maybeRecord(clazz, () -> clazz.getName() + "." + method);
+        return delegate.methodCall(receiver, method, args);
+    }
+
+    @Override public Object constructorCall(Class lhs, Object[] args) throws Throwable {
+        maybeRecord(lhs, () -> lhs.getName() + ".<init>");
+        return delegate.constructorCall(lhs, args);
+    }
+
+    @Override public Object superCall(Class senderType, Object receiver, String method, Object[] args) throws Throwable {
+        maybeRecord(senderType, () -> senderType.getName() + ".<init>");
```

Comment:
I might be missing something, but `<init>` here seems wrong. Not sure if you want to just track it as `senderType.method` or track it against `senderType.getSuperclass()` though.

Reply:
Tried writing some test coverage for a Groovy class extending a Jenkins class but nothing gets reported at all, because we are not using complex logic analogous to `GroovyCallSiteSelector` to determine where a Java member is _defined_; we only check `receiver.class`. So `superCall` is probably useless anyway, though I can change it to report `method` in case it is ever used.